ref(profiles): Count processed profiles with metrics

jjbayer opened the pull request:
As described in https://github.com/getsentry/relay/issues/2158, track the number of processed profiles by tagging the corresponding transaction metric.

Also: Do not accept multiple profiles per envelope.

**Not in this PR:**
* Distinguishing between `Profile` and `ProfileIndexed`
* Rate limiting processed profiles based on metrics

#skip-changelog

jjbayer committed:
feat: Filter profiles

jjbayer committed:
feat: Set has_profile tag

jjbayer committed:
ref: ExtractInput

jjbayer committed:
Merge branch 'ref/extraction-input' into ref/count-profiles-via-metric

jjbayer committed:
Merge branch 'ref/profiles-revert-count' into ref/count-profiles-via-metric

jjbayer committed:
test invalid

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -1061,15 +1065,28 @@ impl EnvelopeProcessorService {
+                    return ItemAction::Drop(Outcome::Invalid(DiscardReason::Profiling(
+                        relay_profiling::discard_reason(ProfileError::TooManyProfiles),
+                    )));
+                }
```
@phacops I added this logic as you suggested.

jjbayer commented on relay-server/src/metrics_extraction/transactions/mod.rs:
```diff
@@ -358,15 +362,21 @@ pub fn extract_transaction_metrics(
     Ok(!added_slice.is_empty())
 }
 
+struct ExtractInput<'a> {
+    aggregator_config: &'a AggregatorConfig,
+    config: &'a TransactionMetricsConfig,
+    event: &'a Event,
+    transaction_from_dsc: Option<&'a str>,
+    sampling_result: &'a SamplingResult,
+    has_profile: bool,
+}
```
The linter finally started complaining about the number of arguments to `extract_transaction_metrics_inner` so I created this struct.

jjbayer committed:
lint

iker-barriocanal commented on relay-server/src/actors/processor.rs:
```diff
@@ -1061,15 +1065,28 @@ impl EnvelopeProcessorService {
+            ItemType::Profile => {
+                if !found_profile {
+                    // Found first profile, keep it.
+                    found_profile = true;
```
Relay also considers towards outcomes dropped profiles that could not parsed, even if they aren't dropped by dynamic sampling. https://github.com/getsentry/relay/pull/2054 mentions only "valid" profiles should be considered:

>In other words, "processed profiles" should count all valid, non rate limited profiles seen by Relays, even if they were dropped by dynamic sampling.

jjbayer replied:
The wording in that PR description is too vague, it should say: "_accepted_ processed profiles should count all valid non rate limited profiles seen by Relays" .

A better way to think of it is two data streams splitting ways (see diagram [here](https://github.com/getsentry/relay/pull/2166)), and getting counted towards a different category regardless of whether they're accepted, rate limited, or invalid.

iker-barriocanal reviewed (approved):
Assuming the PR fulfills the intention in my comment below.

jjbayer committed:
fix: has_profile=false if invalid

jjbayer committed:
feat(profiles): Count as "indexed" after metrics extraction (#2166)

With https://github.com/getsentry/relay/pull/2165, "processed" profiles
are represented by their raw payload _before_ metrics extraction, and
represented by the `transaction.duration` metric _after_ metrics
extraction, as it is the case for transactions:

```mermaid
flowchart LR
    . --"payload\n(DataCategory::Profile)"--> MetricsExtraction
    MetricsExtraction --"payloads\n(TransactionIndexed, ProfileIndexed)"-->..
    MetricsExtraction --"metrics\n(Transaction, Profile)"-->...
 ```


To prevent double counting, emit any profile-related outcome after metrics extraction as `DataCategory::ProfileIndexed`.

Co-authored-by: Iker Barriocanal <32816711+iker-barriocanal@users.noreply.github.com>
